File: java/src/main/java/io/github/xesam/linkseal/resolver/PolicyProcessor.java
package io.github.xesam.linkseal.resolver;

import io.github.xesam.linkseal.model.MissingPolicy;
import io.github.xesam.linkseal.model.PolicyConfig;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class PolicyProcessor {
  public PolicyResult process(Map<String, String> variables, PolicyConfig config) {
    Map<String, String> defaults = config.defaults();
    Map<String, String> result = new LinkedHashMap<>();
    List<String> missing = new ArrayList<>();

    for (Map.Entry<String, String> entry : variables.entrySet()) {
      String key = entry.getKey();
      String value = entry.getValue();
      if (value != null) {
        result.put(key, value);
      } else if (config.missing() == MissingPolicy.DEFAULT && defaults.containsKey(key)) {
        result.put(key, defaults.get(key));
      } else {
        missing.add(key);
      }
    }

    return new PolicyResult(result, missing);
  }

  public Map<String, String> applyMissingPolicy(Map<String, String> variables, PolicyConfig config) {
    PolicyResult processed = process(variables, config);
    if (config.missing() == MissingPolicy.ERROR && !processed.missing().isEmpty()) {
      throw new IllegalArgumentException("Missing required variables: " + String.join(", ", processed.missing()));
    }
    return processed.result();
  }

  public record PolicyResult(Map<String, String> result, List<String> missing) {}
}

File: javascript/packages/linkseal-resolver/src/PolicyProcessor.ts
import type { PolicyConfig } from 'linkseal';

export class PolicyProcessor {
  process(
    variables: Record<string, string | null>,
    config: PolicyConfig
  ): { result: Record<string, string>; missing: string[] } {
    const defaults = config.defaults ?? {};
    const result: Record<string, string> = {};
    const missing: string[] = [];

    for (const [key, value] of Object.entries(variables)) {
      if (value !== null && value !== undefined) {
        result[key] = value;
      } else if (config.missing === 'default' && defaults[key] !== undefined) {
        result[key] = defaults[key];
      } else {
        missing.push(key);
      }
    }

    return { result, missing };
  }

  applyMissingPolicy(
    variables: Record<string, string | null>,
    config: PolicyConfig
  ): { result: Record<string, string>; error?: { code: string; message: string } } {
    const { result, missing } = this.process(variables, config);

    if (config.missing === 'error' && missing.length > 0) {
      return {
        result,
        error: {
          code: 'MISSING_VARIABLE',
          message: `Missing required variables: ${missing.join(', ')}`,
        },
      };
    }

    return { result };
  }
}
